pkg/assembler/backends/inmem/artifact.go

```diff
@@ -81,11 +81,24 @@ func (n *artStruct) setCertifyBadLinks(id uint32) { n.badLinks = append(n.badLin
 // Ingest Artifacts
 
 func (c *demoClient) IngestArtifact(ctx context.Context, artifact *model.ArtifactInputSpec) (*model.Artifact, error) {
+	return c.ingestArtifact(ctx, artifact, true)
+}
+
+func (c *demoClient) ingestArtifact(ctx context.Context, artifact *model.ArtifactInputSpec, readOnly bool) (*model.Artifact, error) {
 	algorithm := strings.ToLower(artifact.Algorithm)
 	digest := strings.ToLower(artifact.Digest)
-	a, err := c.artifactByKey(algorithm, digest)
 
+	c.lock(readOnly)
+	defer c.unlock(readOnly)
+
+	a, err := c.artifactByKey(algorithm, digest)
 	if err != nil {
+		if readOnly {
+			c.m.RUnlock()
```

Comment:
hmm should we always the same interfaces for lock and unlock? (i.e. c.unlock()) ?

Reply:
I think explicitly calling the read unlock is clearer to a future reader to know what is going on (We know we are "readOnly" so release read lock and call-self to write) ?

Reply:
copying notes from our discussion: 

i'd prefer it to be a helper function that isn't a receiver if possible (like lock(c.m,readOnly)instead).

Reply:
Done